File: typescript/src/model/ShoppingCart.ts
import {Product} from "./Product"
import {SupermarketCatalog} from "./SupermarketCatalog"
import * as _ from "lodash"
import {ProductQuantity} from "./ProductQuantity"
import {Discount} from "./Discount"
import {Receipt} from "./Receipt"
import {Offer} from "./Offer"
import {SpecialOfferType} from "./SpecialOfferType"

// Constants for magic numbers
const DEFAULT_ITEM_QUANTITY = 1.0;
const TWO_FOR_AMOUNT_QUANTITY = 2;
const THREE_FOR_TWO_QUANTITY = 3;
const FIVE_FOR_AMOUNT_QUANTITY = 5;
const PERCENTAGE_DIVISOR = 100.0;

// Constants for discount description strings
const THREE_FOR_TWO_DESCRIPTION = "3 for 2";
const TWO_FOR_PREFIX = "2 for ";
const FOR_SEPARATOR = " for ";
const PERCENT_OFF_SUFFIX = "% off";

type ProductQuantities = { [productName: string]: ProductQuantity }
export type OffersByProduct = {[productName: string]: Offer};

export class ShoppingCart {

    private readonly  items: ProductQuantity[] = [];
    _productQuantities: ProductQuantities = {};


    getItems(): ProductQuantity[] {
        return _.clone(this.items);
    }

    addItem(product: Product): void {
        this.addItemQuantity(product, DEFAULT_ITEM_QUANTITY);
    }

    productQuantities(): ProductQuantities {
        return this._productQuantities;
    }


    public addItemQuantity(product: Product, quantity: number): void {
        let productQuantity = new ProductQuantity(product, quantity)
        this.items.push(productQuantity);
        let currentQuantity = this._productQuantities[product.name]
        if (currentQuantity) {
            this._productQuantities[product.name] = this.increaseQuantity(product, currentQuantity, quantity);
        } else {
            this._productQuantities[product.name] = productQuantity;
        }

    }

    private increaseQuantity(product: Product, productQuantity: ProductQuantity, quantity: number) {
        return new ProductQuantity(product, productQuantity.quantity + quantity)
    }

    handleOffers(receipt: Receipt,  offers: OffersByProduct, catalog: SupermarketCatalog ):void {
        for (const productName in this.productQuantities()) {
            const productQuantity = this._productQuantities[productName]
            const product = productQuantity.product;
            const quantity: number = this._productQuantities[productName].quantity;
            if (offers[productName]) {
                const offer : Offer = offers[productName];
                const unitPrice: number= catalog.getUnitPrice(product);
                let quantityAsInt = quantity;
                let discount : Discount|null = null;
                let x = 1;
                if (offer.offerType == SpecialOfferType.ThreeForTwo) {
                    x = THREE_FOR_TWO_QUANTITY;

                } else if (offer.offerType == SpecialOfferType.TwoForAmount) {
                    x = TWO_FOR_AMOUNT_QUANTITY;
                    if (quantityAsInt >= TWO_FOR_AMOUNT_QUANTITY) {
                        const total = offer.argument * Math.floor(quantityAsInt / x) + quantityAsInt % TWO_FOR_AMOUNT_QUANTITY * unitPrice;
                        const discountN = unitPrice * quantity - total;
                        discount = new Discount(product, TWO_FOR_PREFIX + offer.argument, discountN);
                    }

                } if (offer.offerType == SpecialOfferType.FiveForAmount) {
                    x = FIVE_FOR_AMOUNT_QUANTITY;
                }
                const numberOfXs = Math.floor(quantityAsInt / x);
                if (offer.offerType == SpecialOfferType.ThreeForTwo && quantityAsInt > TWO_FOR_AMOUNT_QUANTITY) {
                    const discountAmount = quantity * unitPrice - ((numberOfXs * TWO_FOR_AMOUNT_QUANTITY * unitPrice) + quantityAsInt % THREE_FOR_TWO_QUANTITY * unitPrice);
                    discount = new Discount(product, THREE_FOR_TWO_DESCRIPTION, discountAmount);
                }
                if (offer.offerType == SpecialOfferType.TenPercentDiscount) {
                    discount = new Discount(product, offer.argument + PERCENT_OFF_SUFFIX, quantity * unitPrice * offer.argument / PERCENTAGE_DIVISOR);
                }
                if (offer.offerType == SpecialOfferType.FiveForAmount && quantityAsInt >= FIVE_FOR_AMOUNT_QUANTITY) {
                    const discountTotal = unitPrice * quantity - (offer.argument * numberOfXs + quantityAsInt % FIVE_FOR_AMOUNT_QUANTITY * unitPrice);
                    discount = new Discount(product, x + FOR_SEPARATOR + offer.argument, discountTotal);
                }
                if (discount != null)
                    receipt.addDiscount(discount);
            }

        }
    }
}

File: python_pytest/src/shopping_cart.py
import math

from model_objects import ProductQuantity, SpecialOfferType, Discount

# Constants for magic numbers
DEFAULT_ITEM_QUANTITY = 1.0
TWO_FOR_AMOUNT_QUANTITY = 2
THREE_FOR_TWO_QUANTITY = 3
FIVE_FOR_AMOUNT_QUANTITY = 5
PERCENTAGE_DIVISOR = 100.0

# Constants for discount description strings
THREE_FOR_TWO_DESCRIPTION = "3 for 2"
TWO_FOR_PREFIX = "2 for "
FOR_SEPARATOR = " for "
PERCENT_OFF_SUFFIX = "% off"


class ShoppingCart:

    def __init__(self):
        self._items = []
        self._product_quantities = {}

    @property
    def items(self):
        return self._items

    def add_item(self, product):
        self.add_item_quantity(product, DEFAULT_ITEM_QUANTITY)

    @property
    def product_quantities(self):
        return self._product_quantities

    def add_item_quantity(self, product, quantity):
        self._items.append(ProductQuantity(product, quantity))
        if product in self._product_quantities.keys():
            self._product_quantities[product] = self._product_quantities[product] + quantity
        else:
            self._product_quantities[product] = quantity

    def handle_offers(self, receipt, offers, catalog):
        for p in self._product_quantities.keys():
            quantity = self._product_quantities[p]
            if p in offers.keys():
                offer = offers[p]
                unit_price = catalog.unit_price(p)
                quantity_as_int = int(quantity)
                discount = None
                x = 1
                if offer.offer_type == SpecialOfferType.THREE_FOR_TWO:
                    x = THREE_FOR_TWO_QUANTITY

                elif offer.offer_type == SpecialOfferType.TWO_FOR_AMOUNT:
                    x = TWO_FOR_AMOUNT_QUANTITY
                    if quantity_as_int >= TWO_FOR_AMOUNT_QUANTITY:
                        total = offer.argument * (quantity_as_int / x) + quantity_as_int % TWO_FOR_AMOUNT_QUANTITY * unit_price
                        discount_n = unit_price * quantity - total
                        discount = Discount(p, TWO_FOR_PREFIX + str(offer.argument), -discount_n)

                if offer.offer_type == SpecialOfferType.FIVE_FOR_AMOUNT:
                    x = FIVE_FOR_AMOUNT_QUANTITY

                number_of_x = math.floor(quantity_as_int / x)
                if offer.offer_type == SpecialOfferType.THREE_FOR_TWO and quantity_as_int > TWO_FOR_AMOUNT_QUANTITY:
                    discount_amount = quantity * unit_price - (
                                (number_of_x * TWO_FOR_AMOUNT_QUANTITY * unit_price) + quantity_as_int % THREE_FOR_TWO_QUANTITY * unit_price)
                    discount = Discount(p, THREE_FOR_TWO_DESCRIPTION, -discount_amount)

                if offer.offer_type == SpecialOfferType.TEN_PERCENT_DISCOUNT:
                    discount = Discount(p, str(offer.argument) + PERCENT_OFF_SUFFIX,
                                        -quantity * unit_price * offer.argument / PERCENTAGE_DIVISOR)

                if offer.offer_type == SpecialOfferType.FIVE_FOR_AMOUNT and quantity_as_int >= FIVE_FOR_AMOUNT_QUANTITY:
                    discount_total = unit_price * quantity - (
                                offer.argument * number_of_x + quantity_as_int % FIVE_FOR_AMOUNT_QUANTITY * unit_price)
                    discount = Discount(p, str(x) + FOR_SEPARATOR + str(offer.argument), -discount_total)

                if discount:
                    receipt.add_discount(discount)
